unixfs/src/file/adder.rs

```diff
@@ -0,0 +1,699 @@
+        let expected = [
+            "QmfVyMoStzTvdnUR7Uotzh82gmL427q9z3xW5Y8fUoszi4",
+            "QmdPyW4CWE3QBkgjWfjM5f7Tjb3HukxVuBXZtkqAGwsMnm",
+            "QmNhDQpphvMWhdCzP74taRzXDaEfPGq8vWfFRzD7mEgePM",
+            "Qmc5m94Gu7z62RC8waSKkZUrCCBJPyHbkpmGzEePxy2oXJ",
+            "QmRJHYTNvC3hmd9gJQARxLR1QMEincccBV53bBw524yyq6",
+        ]
+        .iter()
+        .map(|key| {
+            let cid = Cid::try_from(*key).unwrap();
+            let block = blocks.get_by_str(key).to_vec();
+            (cid, block)
+        })
+        .collect::<Vec<_>>();
+
+        assert_eq!(blocks_received, expected);
+    }
+
+    #[test]
+    fn three_layers() {
+        let content = b"Lorem ipsum dolor sit amet, sit enim montes aliquam. Cras non lorem, \
+            rhoncus condimentum, irure et ante. Pulvinar suscipit odio ante, et tellus a enim, \
+            wisi ipsum, vel rhoncus eget faucibus varius, luctus turpis nibh vel odio nulla pede.";
+
+        assert!(content.len() > 174 && content.len() < 2 * 174);
+
+        // go-ipfs 0.5 result: QmRQ6NZNUs4JrCT2y7tmCC1wUhjqYuTssB8VXbbN3rMffg, 239 blocks and root
+        // root has two links:
+        //  - QmXUcuLGKc8SCMEqG4wgct6NKsSRZQfvB2FCfjDow1PfpB (174 links)
+        //  - QmeEn8dxWTzGAFKvyXoLj4oWbh9putL4vSw4uhLXJrSZhs (63 links)
+        //
+        // in future, if we ever add inline Cid generation this test would need to be changed not
+        // to use those inline cids or raw leaves
+        let adder = FileAdder::builder().set_chunker(Chunker::Size(1)).build();
+
+        let blocks_received = adder.collect_blocks(content, 0);
+
+        assert_eq!(blocks_received.len(), 240);
+
+        assert_eq!(
+            blocks_received.last().unwrap().0.to_string(),
+            "QmRQ6NZNUs4JrCT2y7tmCC1wUhjqYuTssB8VXbbN3rMffg"
+        );
+    }
+
+    #[test]
+    fn three_layers_all_subchunks() {
+        let content = b"Lorem ipsum dolor sit amet, sit enim montes aliquam. Cras non lorem, \
+            rhoncus condimentum, irure et ante. Pulvinar suscipit odio ante, et tellus a enim, \
+            wisi ipsum, vel rhoncus eget faucibus varius, luctus turpis nibh vel odio nulla pede.";
+
+        for amt in 1..32 {
+            let adder = FileAdder::builder().set_chunker(Chunker::Size(32)).build();
+            let blocks_received = adder.collect_blocks(content, amt);
+            assert_eq!(
+                blocks_received.last().unwrap().0.to_string(),
+                "QmYSLcVQqxKygiq7x9w1XGYxU29EShB8ZemiaQ8GAAw17h",
+                "amt: {}",
+                amt
+            );
+        }
+    }
+
+    #[test]
+    fn empty_file() {
+        let blocks = FileAdder::default().collect_blocks(b"", 0);
+        assert_eq!(blocks.len(), 1);
+        // 0a == field dag-pb body (unixfs)
+        // 04 == dag-pb body len, varint, 4 bytes
+        // 08 == field type tag, varint, 1 byte
+        // 02 == field type (File)
+        // 18 == field data tag, varlen
```

Comment:
need to recheck this
```suggestion
        // 18 == field filesize tag, varlen
```

Reply:
https://github.com/ribasushi/DAGger/blob/rust_reference/internal/dagger/encoder/unixfsv1/impl.go#L158-L163